torch/distributed/_spmd/graph_optimization.py

```diff
@@ -548,3 +554,339 @@ def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+            if nodes:
+                nodes.append(None)
+            continue
+        if node.op == OP.CALL_FUNCTION and str(node.target).startswith(
+            "aten._foreach_add"
+        ):
+            step_node = node
+            break
+        else:
+            flatten_args, _ = tree_flatten((node.args, node.kwargs))
+            for ancestor in flatten_args:
+                if isinstance(ancestor, fx.Node):
+                    nodes.append(ancestor)
+    if step_node == optim_node:
+        raise RuntimeError(
+            "Cannot find step node (foreach_add) for the optimizer node "
+            f"{optim_node}. The API design does not match the tracing graph."
+        )
+
+    step = ForeachAddBlock(step_node, generate_output=False)
+    optim = FusedAdamBlock(optim_node, generate_output=False)
+    return FusedOptimizerBlock(step, optim)
+
+
+def get_all_fused_optimizer_blocks(
+    gm: IterGraphModule, optim_ops: Union[Tuple[str, ...], str]
+) -> List[FusedOptimizerBlock]:
+    """
+    Find all the FusedOptimizerBlock that the optimizer operators are in
+    `optim_ops`.
+    """
+    return [
+        get_fused_optimizer_block(node)
+        for node in gm.graph.nodes
+        if node.name.startswith(optim_ops)
+    ]
+
+
+def _split_fused_adam(
+    gm: IterGraphModule,
+    orig_optim_block: FusedOptimizerBlock,
+    split_gradients: Set[fx.Node],
+) -> Tuple[FusedOptimizerBlock, FusedOptimizerBlock]:
+    """
+    Split the `orig_optim_block` into two FusedOptimizerBlock. The first one
+    will be the optimizer that optimize `split_gradients`. The second one is
+    used to optimize the remaining gradients.
+    An assert will be raised if one of the optimizer optimize zero gradients.
+    """
+    orig_optim_args = AdamArgs(*orig_optim_block.optim.optim_node.args)
+    optim_args = (AdamArgs([], [], [], [], [], []), AdamArgs([], [], [], [], [], []))
+    # The only hint we can use to split the optimizer is the order/indices.
+    orig_optim_indices: Tuple[List[int], List[int]] = ([], [])
+    orig_step_indices: Tuple[List[int], List[int]] = ([], [])
+
+    for idx, gradient in enumerate(orig_optim_args.grads):
+        group_idx = 0 if gradient in split_gradients else 1
+        orig_optim_indices[group_idx].append(idx)
+        # Get the argument for idx-th gradient from orig_optim_args
+        for orig_arg, optim_arg in zip(orig_optim_args, optim_args[group_idx]):
+            # Only add the argument to the list if the original argument list
+            # is not empty. If the original argument list is empty, the new
+            # one must be an empty list as well.
+            if orig_arg:
+                optim_arg.append(orig_arg[idx])
+
+        # If argument order of step is the same as optimizer, nothing has to be
+        # done. However, it is risky to rely on this assumption so we populate
+        # the orig_step_indices.
+        old_step_output = optim_args[group_idx].state_steps[-1]
```

Comment:
what's the difference between old vs orig prefix? Or should this be `orig_step_output`?

Reply:
Good catch. Will change it.